File: app/src/main/java/com/android/example/popularmovies/ui/detail/fragments/videos/VideosFragment.java
package com.android.example.popularmovies.ui.detail.fragments.videos;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.android.example.popularmovies.LoadingStatus;
import com.android.example.popularmovies.R;
import com.android.example.popularmovies.database.FavMovieEntry;
import com.android.example.popularmovies.databinding.FragmentVideosBinding;
import com.android.example.popularmovies.ui.detail.fragments.BaseDetailFragment;
import com.android.example.popularmovies.utils.json.MovieVideo;

public class VideosFragment extends BaseDetailFragment implements VideosAdapter.VideoOnClickListener {

    private VideosViewModel viewModel;
    private FragmentVideosBinding binding;
    private VideosAdapter adapter;

    public static VideosFragment newInstance() {
        return new VideosFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentVideosBinding.inflate(inflater, container, false);

        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext());
        binding.rvVideoList.setLayoutManager(layoutManager);

        adapter = new VideosAdapter(this);
        binding.rvVideoList.setAdapter(adapter);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(VideosViewModel.class);
        viewModel.getStatus().observe(getViewLifecycleOwner(), status -> {
            switch (status){
                case ERROR:
                    binding.tvErrorText.setVisibility(View.VISIBLE);
                    binding.llVideoLayout.setVisibility(View.INVISIBLE);
                    binding.progressBar.setVisibility(View.INVISIBLE);
                    break;
                case LOADING:
                    binding.progressBar.setVisibility(View.VISIBLE);
                    binding.tvErrorText.setVisibility(View.INVISIBLE);
                    binding.llVideoLayout.setVisibility(View.INVISIBLE);
                    break;
                case DONE:
                    binding.llVideoLayout.setVisibility(View.VISIBLE);
                    binding.tvErrorText.setVisibility(View.INVISIBLE);
                    binding.progressBar.setVisibility(View.INVISIBLE);
                    break;
            }
        });
        viewModel.getVideos().observe(getViewLifecycleOwner(), movieVideos -> {
            if (movieVideos != null && !movieVideos.isEmpty()){
                adapter.setVideos(movieVideos);
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        FavMovieEntry movie = getMovie(); //TODO: Move to onViewCreated?
        if (movie != null) {
            viewModel.fetchVideos(getString(R.string.api_key_v3), movie.getId());
        } else {
            viewModel.setStatus(LoadingStatus.ERROR);
            Log.e(getClass().getSimpleName(), "Null movie passed to fragment.");
        }
    }

    @Override
    public void onListItemClick(MovieVideo video) {
        Intent intent = new Intent(Intent.ACTION_VIEW, video.getVideoUri())
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        if (intent.resolveActivity(requireActivity().getPackageManager()) != null) {
            startActivity(intent);
        }
    }
}

File: app/src/main/java/com/android/example/popularmovies/ui/detail/fragments/videos/VideosFragment.kt
package com.android.example.popularmovies.ui.detail.fragments.videos

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.android.example.popularmovies.R
import com.android.example.popularmovies.databinding.FragmentVideosBinding
import com.android.example.popularmovies.ui.detail.fragments.BaseDetailFragment
import com.android.example.popularmovies.ui.detail.fragments.videos.VideosAdapter.VideoOnClickListener
import com.android.example.popularmovies.utils.json.MovieVideo
import com.android.example.popularmovies.utils.network.LoadingStatus

class VideosFragment : BaseDetailFragment(), VideoOnClickListener {
    private val viewModel by viewModels<VideosViewModel>()
    private lateinit var binding: FragmentVideosBinding
    private lateinit var adapter: VideosAdapter
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentVideosBinding.inflate(inflater, container, false)
        val layoutManager = LinearLayoutManager(requireContext())
        binding.rvVideoList.layoutManager = layoutManager
        adapter = VideosAdapter(this)
        binding.rvVideoList.adapter = adapter
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.status.observe(viewLifecycleOwner, Observer { status: LoadingStatus? ->
            when (status) {
                LoadingStatus.ERROR -> {
                    binding.tvErrorText.visibility = View.VISIBLE
                    binding.llVideoLayout.visibility = View.INVISIBLE
                    binding.progressBar.visibility = View.INVISIBLE
                }
                LoadingStatus.LOADING -> {
                    binding.progressBar.visibility = View.VISIBLE
                    binding.tvErrorText.visibility = View.INVISIBLE
                    binding.llVideoLayout.visibility = View.INVISIBLE
                }
                LoadingStatus.DONE -> {
                    binding.llVideoLayout.visibility = View.VISIBLE
                    binding.tvErrorText.visibility = View.INVISIBLE
                    binding.progressBar.visibility = View.INVISIBLE
                }
            }
        })
        viewModel!!.videos.observe(viewLifecycleOwner, Observer { movieVideos: List<MovieVideo>? ->
            if (movieVideos != null && movieVideos.isNotEmpty()) {
                adapter.setVideos(movieVideos)
            }
        })
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val movie = movie
        if (movie != null) {
            viewModel.fetchVideos(getString(R.string.api_key_v3), movie.id)
        } else {
            viewModel.setStatus(LoadingStatus.ERROR)
            Log.e(javaClass.simpleName, "Null movie passed to fragment.")
        }
    }

    override fun onListItemClick(video: MovieVideo) {
        val intent = Intent(Intent.ACTION_VIEW, video.videoUri)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (intent.resolveActivity(requireActivity().packageManager) != null) {
            startActivity(intent)
        }
    }

    companion object {
        fun newInstance(): VideosFragment {
            return VideosFragment()
        }
    }
}